### ISSUE TITLE
Invalid `E0308` errors

### ISSUE THREAD

**Issue by stinodego** [rust-lang/rust-analyzer#16843]
**rust-analyzer version**: (eg. output of "rust-analyzer: Show RA Version" command, accessible in VSCode via <kbd>Ctrl/⌘</kbd>+<kbd>Shift</kbd>+<kbd>P</kbd>)

**rustc version**: (eg. output of `rustc -V`)

**relevant settings**: (eg. client settings, or environment variables like `CARGO`, `RUSTC`, `RUSTUP_HOME` or `CARGO_HOME`)

**repository link (if public, optional)**: (eg. [rust-analyzer](https://github.com/rust-lang/rust-analyzer))

**code snippet to reproduce**: 
```rust
// add your code here

```

I couldn't manage to create a small reproducible example, but it should be easy to reproduce by cloning our repo at this commit:
https://github.com/pola-rs/polars/tree/bb1aa267e92078a3e69bec6d94a4f8b5b8d840c7

The issue is with the following file:
https://github.com/pola-rs/polars/blob/bb1aa267e92078a3e69bec6d94a4f8b5b8d840c7/crates/polars-core/src/series/any_value.rs

rust-analyzer finds two problems (E0308). However, when fixing these problems, the code no longer compiles.

I am using rust-analyzer `v0.3.1877` with VSCode on WSL2 with Ubuntu 22.02.

I hope this report helps address the issue. I am happy to provide more details if necessary.
